include/posix/pthread.h

```diff
@@ -7,38 +7,58 @@
 #ifndef __PTHREAD_H__
 #define __PTHREAD_H__
 
-#ifdef CONFIG_NEWLIB_LIBC
+#include <kernel.h>
 #include <time.h>
-#else
-/* This should probably live somewhere else but Zephyr doesn't
- * currently have a stdc layer to provide it
- */
-struct timespec {
-	s32_t tv_sec;
-	s32_t tv_nsec;
+#include "sys/types.h"
+#include "posix_sched.h"
+#include "unistd.h"
+
+/* Thread state. */
+enum pthread_state {
+	/* The thread is running and joinable. */
+	PTHREAD_JOINABLE = 0,
+	/* The thread is running and detached. */
+	PTHREAD_DETACHED,
+	/* A joinable thread exited and its return code is available. */
+	PTHREAD_EXITED,
+	/* The thread structure is unallocated and available for reuse. */
+	PTHREAD_TERMINATED
 };
-#endif
 
-static inline s32_t _ts_to_ms(const struct timespec *to)
-{
-	return (to->tv_sec * 1000) + (to->tv_nsec / 1000000);
-}
+struct posix_thread {
+	struct k_thread thread;
+
+	/* Exit status */
+	void *retval;
 
-typedef struct pthread_mutex {
-	struct k_sem *sem;
-} pthread_mutex_t;
+	/* Pthread cancellation */
+	int cancel_state;
+	int cancel_pending;
+	struct k_sem cancel_lock_sem;
+	pthread_mutex_t cancel_lock;
```

Comment:
Why do we have a semaphore and a pthread_mutex?

Reply:
pthread_mutex_t just contains a pointer to a semphore. 
So we need a semaphore to for this.

Reply:
A user using these APIs will have to define a k_sem as well?

Reply:
pthread mutex was implemented by @andyross in last zephyr release.

As per my understanding, there are 2 ways to achieve this :- 
1. Application need to define it statically something like :  "PTHREAD_MUTEX_DEFINE(lock);" (something shown in same application tests/kernel/pthread)
2. Application need to define a "k_sem" and then use ptherad_mutex_init to initialize it. (approach I have taken for my implementation for achieving cancel_lock and state_lock).

I think we need to modify this later, as it does not seem intuitive to a app developer.

Reply:
Can we make it work for both cases (by statically initializing it with that macro or calling pthread_init()) if pthread_mutex_t were embedding a k_sem rather than pointing to it?

Reply:
> I think we need to modify this later

I think we need to fix this now.

Reply:
Yes @andrewboie we need to fix this now. But it will need changes in almost all mutex related APIs. At this stage I am avoiding this because merge window is approaching and this will gate all POSIX patches. I will put a FIXME: Comment here and will take this up after 1.11.